File: autograder/services/templates/notebook_template.rb
require 'json'
require 'base64'
require 'stringio'
require 'timeout'

CELLS_B64 = "{cells_b64}"
TEST_CODE_B64 = "{test_code_b64}"

def capture_output
  old_stdout = $stdout
  old_stderr = $stderr
  out = StringIO.new
  err = StringIO.new
  $stdout = out
  $stderr = err
  yield
  return out.string, err.string
ensure
  $stdout = old_stdout
  $stderr = old_stderr
end

cells_json = Base64.decode64(CELLS_B64)
cells = JSON.parse(cells_json)

results = []
execution_count = 0
# ============================================
# Shared Execution Context
# ============================================
$execution_context = Object.new

cells.each do |cell|
  if cell['type'] == 'markdown'
    results << {
      "cell_type" => "markdown",
      "source" => cell['source']
    }
  elsif cell['type'] == 'code'
    execution_count += 1
    source = cell['source']
    # Handle source as array or string
    source = source.join("") if source.is_a?(Array)
    outputs = []
    success = true
    result_val = nil
    error_obj = nil
    
    stdout_str, stderr_str = capture_output do
      begin
        # Execute in shared context
        result_val = $execution_context.instance_eval(source)
      rescue Exception => e
        success = false
        error_obj = e
        # Print error to stderr so we catch it
        $stderr.puts e.message
        $stderr.puts e.backtrace.join("\n")
      end
    end

    if !stdout_str.empty?
      outputs << {
        "output_type" => "stream",
        "name" => "stdout",
        "text" => stdout_str
      }
    end

    if !stderr_str.empty?
       # If failed, stderr usually contains trace
       if success
          outputs << {
            "output_type" => "stream",
            "name" => "stderr",
            "text" => stderr_str
          }
       end
    end

    if !success
      outputs << {
        "output_type" => "error",
        "ename" => error_obj.class.to_s,
        "evalue" => error_obj.message,
        "traceback" => error_obj.backtrace || []
      }
    elsif result_val != nil
       outputs << {
         "output_type" => "execute_result",
         "execution_count" => execution_count,
         "data" => {
           "text/plain" => result_val.inspect
         },
         "metadata" => {}
       }
    end

    results << {
      "cell_type" => "code",
      "source" => source,
      "outputs" => outputs,
      "execution_count" => execution_count
    }
  end
end

# ============================================
# Tester Framework for Ruby Notebooks
# ============================================
$test_results = []

# Define run_test at top level so it is a private method of Object
# instance_eval will follow inheritance chain and find it? 
# Actually, instance_eval sets self. calling run_test matches self.run_test.
# Since run_test is private on Object, and self is an Object, it works.
def run_test(name, points, description=nil, timeout=30, &block)
  if description.is_a?(Proc)
    block = description
    description = nil
  end
  if timeout.is_a?(Proc)
    block = timeout
    timeout = 30
  end

  result = {
    "name" => name,
    "max_score" => points,
    "description" => description,
    "message" => "",
    "score" => 0,
    "passed" => false,
    "status" => "failed",
    "error" => ""
  }
  
  begin
    ret = nil
    Timeout.timeout(timeout) do
      ret = block.call
    end

    if ret.is_a?(Numeric)
      result["score"] = [[ret, points].min, 0].max
      result["passed"] = result["score"] == points
      result["status"] = result["passed"] ? "passed" : "partial"
    elsif ret.is_a?(Array) && ret.length >= 2 && ret[0].is_a?(Numeric)
      result["score"] = [[ret[0], points].min, 0].max
      result["message"] = ret[1].nil? ? "" : ret[1].to_s
      result["passed"] = result["score"] == points
      result["status"] = result["passed"] ? "passed" : "partial"
    elsif ret.is_a?(String)
      result["message"] = ret
      result["passed"] = true
      result["score"] = points
      result["status"] = "passed"
    else
      result["passed"] = true
      result["score"] = points
      result["status"] = "passed"
    end
  rescue Exception => e
    result["error"] = e.message
    result["status"] = e.is_a?(Timeout::Error) ? "error" : "failed"
  end
  
  $test_results << result
end

def output_test_results
  puts "<<<TEST_RESULT_JSON_START>>>"
  puts JSON.generate($test_results)
  puts "<<<TEST_RESULT_JSON_END>>>"
end

# Make run_test public just in case instance_eval scope is tricky with private methods
public :run_test

# ============================================

# Execute test code if provided
if !TEST_CODE_B64.empty?
  begin
    test_code = Base64.decode64(TEST_CODE_B64)
    if !test_code.strip.empty?
      # Execute test code in the same context
      $execution_context.instance_eval(test_code)
    end
  rescue Exception => e
    # Manually create a failed result if the script itself blows up
    $test_results << {
        "name" => "Test Script Execution",
        "max_score" => 0,
        "score" => 0,
        "passed" => false,
        "status" => "failed",
        "error" => "Failed to run test script: #{e.message}"
    }
  end
end

# Output test results
output_test_results

final_result = {
  "success" => true,
  "stdout" => "",
  "stderr" => "",
  "execution_time" => 0,
  "output_data" => {
    "cells" => results,
    "notebook" => ""
  }
}

puts "<<<RESULTS_START>>>"
puts JSON.generate(final_result)
puts "<<<RESULTS_END>>>"

File: autograder/services/templates/notebook_template.r
# This is a template for running R Jupyter notebook code cells inside a Docker container.

# To use this template replace the placeholder {cells_b64} with a base64-encoded JSON array of cells, And packages_to_install with a list of packages to install.

MAX_CELLS <- 500  # Maximum number of cells allowed to prevent abuse

packages_to_install <- list()

# template_log is a function that logs the template to a file.
template_log <- function(message) {
   cat(paste0(message, "\n"), file = stderr())
}

file_exists <- function(file_path) {
   file.info(file_path)$size > 0
}
dir_exists <- function(dir_path) {
   file.info(dir_path)$isdir
}

template_log("Template started.")
start_time <- Sys.time()
template_log(paste("Initial working directory: ", getwd()))
template_log(paste("/work exists: ", dir_exists("/work")))
template_log(paste("/root/shared exists: ", dir_exists("/root/shared")))

# if /work exists, set it as the working directory
if (dir_exists("/work")) {
   setwd("/work")
   template_log(paste("Working directory set to: ", getwd()))
} else {
   template_log("/work directory does not exist.")
   q(save="no", status=1)
}

# Install required packages if not present
required_pkgs <- c("base64enc", "jsonlite")
new_pkgs <- required_pkgs[!(required_pkgs %in% installed.packages()[,"Package"])]
if(length(new_pkgs)) {
    template_log(paste("Installing required packages:", paste(new_pkgs, collapse = ", ")))
    install.packages(new_pkgs, repos = "https://cloud.r-project.org")
}

library(base64enc)
library(jsonlite)


# install packages
if (length(packages_to_install) > 0) {
   install.packages(packages_to_install)
   template_log(paste("Packages installed: ", paste(packages_to_install, collapse = ", ")))
   # Log to stderr (using message or cat(file=stderr())) so Converger picks it up
   for (pkg in packages_to_install) {
       message(paste0("CODEPOST_AUTO_INSTALL_SUCCESS: ", pkg))
   }
} else {
   template_log("No packages to install.")
}

# Security: Make cache read-only after installation to prevent notebook code from tampering
if (length(packages_to_install) > 0) {
   template_log("Making cache read-only after installation.")
   # Attempt to lock standard cache locations if they exist and are writable
   cache_dirs <- c(Sys.getenv("R_LIBS_USER"), "/tmp/pip-cache", "/tmp/npm-cache") 
   for (cache_dir in cache_dirs) {
       if (dir.exists(cache_dir)) {
           tryCatch({
               Sys.chmod(cache_dir, mode = "0555", use_umask = FALSE)
           }, error = function(e) {
               template_log(paste("Could not lock cache dir:", cache_dir))
           })
       }
   }
}

# flush console
flush.console()

# decode cells
decoded_raw <- base64decode('{cells_b64}')
decoded <- rawToChar(decoded_raw)

# parse cells - use simplifyVector=FALSE to get proper list structure
notebook_json <- fromJSON(decoded, simplifyVector = FALSE)

# check if cells are valid - use 'type' key as that's what prepare_notebook uses
if (!is.list(notebook_json)) {
    template_log("notebook_json is not a list")
    q(save="no", status=1)
}
results <- list()

# Use global environment for cell execution (like Jupyter R kernels do)
cell_env <- globalenv()

# check if cells are too many
if (length(notebook_json) > MAX_CELLS) {
   template_log(paste("Too many cells: ", length(notebook_json)))
   results <- list(
    type='markdown',
    source=paste("**Error:** Too many cells. Maximum allowed: ", MAX_CELLS, "\n\n", "Execution stopped.")
   )
} else {
    NBS_OUTPUT_LIMIT <- 10000  # Max characters in output 10kb
    execution_count <- 0
    
    for (cell in notebook_json) {
        if (cell$type == "markdown") {
            results <- c(results, list(list(
                cell_type = 'markdown',
                source = cell$source
            )))
        }
        if (cell$type == "code") {
            execution_count <- execution_count + 1
            # Handle source as array or string
            cell_source <- if (is.list(cell$source) || is.character(cell$source) && length(cell$source) > 1) {
                paste(unlist(cell$source), collapse = "")
            } else {
                cell$source
            }
            
            outputs <- list()
            success <- TRUE
            error_msg <- NULL
            
            # Capture stdout
            stdout_file <- tempfile(fileext = ".txt")
            stderr_file <- tempfile(fileext = ".txt")
            
            # Create a temp file for potential plot output
            plot_file <- tempfile(fileext = ".png")
            
            # Parse and eval the code BEFORE any output capture
            # R's sink() interferes with eval() when capturing output
            parsed <- tryCatch({
                parse(text = cell_source)
            }, error = function(e) {
                success <<- FALSE
                error_msg <<- conditionMessage(e)
                NULL
            })
            
            if (!is.null(parsed) && length(parsed) > 0) {
                for (expr in parsed) {
                    tryCatch({
                        eval(expr, envir = globalenv())
                    }, error = function(e) {
                        success <<- FALSE
                        error_msg <<- conditionMessage(e)
                    })
                }
            }
            
            tryCatch({
                # Open graphics device to capture plots
                png(plot_file, width = 800, height = 600, res = 100)
                
                # Capture stdout and stderr
                sink(stdout_file, type = "output")
                sink(stderr_file, type = "message")
                
                # Parse and evaluate the code - already done above, just need output
                if (!is.null(parsed) && length(parsed) > 0) {
                    last_value <- NULL
                    # Directly evaluate the parsed code in global environment
                    for (i in seq_along(parsed)) {
                        expr <- parsed[[i]]
                        last_value <- tryCatch({
                            eval(expr)
                        }, error = function(e) {
                            success <<- FALSE
                            error_msg <<- conditionMessage(e)
                            NULL
                        })
                    }
                }
                
                
            }, error = function(e) {
                success <<- FALSE
                error_msg <<- conditionMessage(e)
            }, warning = function(w) {
                # Capture warnings but continue
                cat(paste("Warning:", conditionMessage(w), "\n"), file = stderr_file, append = TRUE)
                invokeRestart("muffleWarning")
            }, finally = {
                # Close sinks
                sink(type = "message")
                sink(type = "output")
                
                # Close graphics device
                dev.off()
            })
            
            # Read captured stdout
            stdout_text <- ""
            if (file.exists(stdout_file) && file.info(stdout_file)$size > 0) {
                stdout_text <- paste(readLines(stdout_file, warn = FALSE), collapse = "\n")
            }
            
            # Read captured stderr
            stderr_text <- ""
            if (file.exists(stderr_file) && file.info(stderr_file)$size > 0) {
                stderr_text <- paste(readLines(stderr_file, warn = FALSE), collapse = "\n")
            }
            
            # Check if a plot was created
            # Note: Empty PNG files from R are typically ~100-200 bytes
            # A real plot will be at least 1KB, so we use 1000 bytes as threshold
            EMPTY_PLOT_THRESHOLD <- 1000
            if (file.exists(plot_file) && file.info(plot_file)$size > EMPTY_PLOT_THRESHOLD) {
                # File is large enough to contain actual plot content
                tryCatch({
                    # Read plot and encode as base64
                    plot_data <- readBin(plot_file, "raw", file.info(plot_file)$size)
                    plot_base64 <- base64encode(plot_data)
                    
                    outputs <- c(outputs, list(list(
                        output_type = 'display_data',
                        data = list(
                            `image/png` = plot_base64
                        ),
                        metadata = list()
                    )))
                }, error = function(e) {
                    # If plot capture fails, add to stderr
                    stderr_text <- paste0(stderr_text, "\nPlot capture error: ", conditionMessage(e), "\n")
                })
            }
            
            # Clean up temp files
            unlink(c(stdout_file, stderr_file, plot_file))
            
            # Truncate stdout if too long
            if (nchar(stdout_text) > NBS_OUTPUT_LIMIT) {
                stdout_text <- paste0(substr(stdout_text, 1, NBS_OUTPUT_LIMIT), "\n...[ERROR output truncated]\n")
            }
            
            # Truncate stderr if too long
            if (nchar(stderr_text) > NBS_OUTPUT_LIMIT) {
                stderr_text <- paste0(substr(stderr_text, 1, NBS_OUTPUT_LIMIT), "\n...[ERROR output truncated]\n")
            }
            
            # Build outputs list
            if (nchar(stdout_text) > 0) {
                outputs <- c(outputs, list(list(
                    output_type = 'stream',
                    name = 'stdout',
                    text = stdout_text
                )))
            }
            
            if (nchar(stderr_text) > 0) {
                outputs <- c(outputs, list(list(
                    output_type = 'stream',
                    name = 'stderr',
                    text = stderr_text
                )))
            }
            
            if (!success) {
                # Truncate error message if too long
                if (!is.null(error_msg) && nchar(error_msg) > NBS_OUTPUT_LIMIT) {
                    error_msg <- paste0(substr(error_msg, 1, NBS_OUTPUT_LIMIT), "\n...[ERROR message truncated]\n")
                }
                
                outputs <- c(outputs, list(list(
                    output_type = 'error',
                    ename = 'ExecutionError',
                    evalue = if (!is.null(error_msg)) error_msg else stderr_text,
                    traceback = list(stderr_text)
                )))
            }
            
            results <- c(results, list(list(
                cell_type = 'code',
                source = cell_source,
                outputs = outputs,
                execution_count = execution_count
            )))
        }
    }
}

# Output results as JSON
end_time <- Sys.time()
execution_time <- as.numeric(difftime(end_time, start_time, units = "secs"))

# ============= Tester Framework =============
test_results <- list()

run_test <- function(name, points, description, fn = NULL, timeout = 30) {
    if (is.function(description)) {
        timeout <- fn
        fn <- description
        description <- NULL
    }

    result <- list(
        name = name,
        max_score = points,
        description = description,
        message = "",
        score = 0,
        passed = FALSE,
        status = "failed",
        error = ""
    )

    tryCatch({
        # Re-assign fn's environment to cell_env so it can access notebook functions
        environment(fn) <- cell_env
        setTimeLimit(elapsed = timeout, transient = TRUE)
        on.exit(setTimeLimit(elapsed = Inf, transient = TRUE), add = TRUE)

        ret <- fn()
        if (is.numeric(ret) && length(ret) >= 1) {
            score <- as.numeric(ret[[1]])
            result$score <- max(0, min(score, points))
            result$passed <- result$score == points
            result$status <- if (result$passed) "passed" else "partial"
        } else if (is.list(ret) && length(ret) >= 2 && is.numeric(ret[[1]])) {
            score <- as.numeric(ret[[1]])
            result$score <- max(0, min(score, points))
            result$message <- if (!is.null(ret[[2]])) as.character(ret[[2]]) else ""
            result$passed <- result$score == points
            result$status <- if (result$passed) "passed" else "partial"
        } else if (is.list(ret) && !is.null(ret$score) && is.numeric(ret$score)) {
            score <- as.numeric(ret$score)
            result$score <- max(0, min(score, points))
            result$message <- if (!is.null(ret$message)) as.character(ret$message) else ""
            result$passed <- result$score == points
            result$status <- if (result$passed) "passed" else "partial"
        } else if (is.logical(ret) && length(ret) >= 1) {
            result$passed <- as.logical(ret[[1]])
            result$score <- if (result$passed) points else 0
            result$status <- if (result$passed) "passed" else "failed"
        } else if (is.character(ret) && length(ret) >= 1) {
            result$message <- as.character(ret[[1]])
            result$passed <- TRUE
            result$score <- points
            result$status <- "passed"
        } else {
            result$passed <- TRUE
            result$score <- points
            result$status <- "passed"
        }
    }, error = function(e) {
        result$error <<- conditionMessage(e)
        result$status <<- if (grepl("time limit|timed out|timeout", tolower(conditionMessage(e)))) "error" else "failed"
    })

    test_results <<- c(test_results, list(result))
}

output_test_results <- function() {
    cat('<<<TEST_RESULT_JSON_START>>>')
    cat(toJSON(test_results, auto_unbox = TRUE))
    cat('<<<TEST_RESULT_JSON_END>>>')
}
# ============================================

# Execute test code if provided
test_code_b64 <- '{test_code_b64}'
if (nchar(test_code_b64) > 0) {
    tryCatch({
        test_code <- rawToChar(base64decode(test_code_b64))
        if (nchar(trimws(test_code)) > 0) {
            eval(parse(text = test_code), envir = cell_env)
        }
    }, error = function(e) {
        run_test("Test Script Execution", 0, function() {
            stop(paste("Failed to run test script:", conditionMessage(e)))
        })
    })
}

# Output test results
output_test_results()

final_result <- list(
    success = TRUE, # If we got here without crashing, we consider it a success at the notebook level
    stdout = "",
    stderr = "",
    error = NULL,
    execution_time = execution_time,
    output_data = list(
        cells = results
    )
)

cat('<<<RESULTS_START>>>\n')
cat(toJSON(final_result, auto_unbox = TRUE, null = "null"))
cat('\n<<<RESULTS_END>>>\n')


